### PR TITLE
Crystal 0.35.1 updates

### PR DESCRIPTION
**Pull request by rmarronnier**
This PR takes care of deprecated `JSON.Mapping`, `Time.new`, and `Logger`

Note : Logs are not displayed when running crystal specs with the new `Crystal::Log` (but logs are displayed when using Shainet normally)

### CODE REVIEW

File: src/shainet/basic/layer.cr [hunk 1]
```diff
@@ -3,12 +3,13 @@ require "apatite"
 module SHAInet
   class Layer
     include Apatite
+    LOG = Log.for("Layer")
```
Comment (Sija) [inline src/shainet/basic/layer.cr:6]:
```suggestion
    Log = ::Log.for(self)
```

File: src/shainet/data/data.cr [hunk 2]
```diff
@@ -2,6 +2,8 @@ require "csv"
 
 module SHAInet
   class Data
+    LOG = Log.for("Data")
```
Comment (Sija) [inline src/shainet/data/data.cr:5]:
```suggestion
    Log = ::Log.for(self)
```

File: src/shainet/basic/network_setup.cr [hunk 3]
```diff
@@ -14,6 +14,8 @@ module SHAInet
     # the network, for methods regarding running and training go to network_run.cr
     # ------------
 
+    LOG = Log.for("Network")
```
Comment (Sija) [inline src/shainet/basic/network_setup.cr:17]:
```suggestion
    Log = ::Log.for(self)
```

File: src/shainet/cnn/cnn.cr [hunk 4]
```diff
@@ -9,6 +9,8 @@ module SHAInet
   # Structure hierarchy: CNN > Layer > Filter > Channel > Row > Neuron/Synapse
 
   class CNN
+    LOG = Log.for("CNN")
```
Comment (Sija) [inline src/shainet/cnn/cnn.cr:12]:
```suggestion
    Log = ::Log.for(self)
```

File: src/shainet/cnn/cnn_input_layer.cr [hunk 5]
```diff
@@ -1,10 +1,12 @@
-require "logger"
+require "log"
 
 module SHAInet
   class InputLayer
+    LOG = Log.for("InputLayer")
```
Comment (Sija) [inline src/shainet/cnn/cnn_input_layer.cr:5]:
```suggestion
    Log = ::Log.for(self)
```

File: src/shainet/cnn/conv_layer.cr [hunk 6]
```diff
@@ -1,7 +1,9 @@
-require "logger"
+require "log"
 
 module SHAInet
   class ConvLayer
+    LOG = Log.for("ConvLayer")
```
Comment (Sija) [inline src/shainet/cnn/conv_layer.cr:5]:
```suggestion
    Log = ::Log.for(self)
```

File: src/shainet/cnn/drop_out_layer.cr [hunk 7]
```diff
@@ -1,13 +1,14 @@
-require "logger"
+require "log"
 
 module SHAInet
   class DropoutLayer
+    LOG = Log.for("DropoutLayer")
```
Comment (Sija) [inline src/shainet/cnn/drop_out_layer.cr:5]:
```suggestion
    Log = ::Log.for(self)
```

File: src/shainet/cnn/fc_layer.cr [hunk 8]
```diff
@@ -1,15 +1,15 @@
-require "logger"
+require "log"
 
 module SHAInet
   class FullyConnectedLayer
+    LOG = Log.for("FullyConnectedLayer")
```
Comment (Sija) [inline src/shainet/cnn/fc_layer.cr:5]:
```suggestion
    Log = ::Log.for(self)

```

File: src/shainet/cnn/max_pool_layer.cr [hunk 9]
```diff
@@ -1,13 +1,13 @@
-require "logger"
+require "log"
 
 module SHAInet
   class MaxPoolLayer
+    LOG = Log.for("MaxPoolLayer")
```
Comment (Sija) [inline src/shainet/cnn/max_pool_layer.cr:5]:
```suggestion
    Log = ::Log.for(self)

```

File: src/shainet/cnn/relu_layer.cr [hunk 10]
```diff
@@ -1,13 +1,13 @@
-require "logger"
+require "log"
 
 module SHAInet
   class ReluLayer
+    LOG = Log.for("ReluLayer")
```
Comment (Sija) [inline src/shainet/cnn/relu_layer.cr:5]:
```suggestion
    Log = ::Log.for(self)

```

File: src/shainet/cnn/softmax_layer.cr [hunk 11]
```diff
@@ -1,11 +1,12 @@
-require "logger"
+require "log"
 
 module SHAInet
   class SoftmaxLayer
+    LOG = Log.for("SoftmaxLayer")
```
Comment (Sija) [inline src/shainet/cnn/softmax_layer.cr:5]:
```suggestion
    Log = ::Log.for(self)

```

File: src/shainet/basic/layer.cr (hunk 1 above)
Comment (Sija) [inline src/shainet/basic/layer.cr:6]:
```suggestion

    Log = ::Log.for(self)

```

File: src/shainet/rnn/rnn.cr [hunk 12]
```diff
@@ -10,6 +10,7 @@ module SHAInet
 
   # Recurrent Neural Network
   class RNN
+    LOG            = Log.for("RNN")
```
Comment (Sija) [inline src/shainet/rnn/rnn.cr:13]:
```suggestion
    Log            = ::Log.for(self)
```

File: src/shainet/basic/exprimental.cr [hunk 13]
```diff
@@ -39,7 +39,7 @@ module SHAInet
       end
 
       unless stealth # Hide report during training
-        @input_layers.last.activations.show(@logger, "input layer activations:")
+        @input_layers.last.activations.show(LOG, "input layer activations:")
```
Comment (Sija) [inline src/shainet/basic/exprimental.cr:42]: I don't think there's a need to pass a `Log` instance here anymore.
ditto elsewhere

**Comment by rmarronnier** [comment]
Thanks for the review !

Ameba is not happy with the `Log` constant naming though : `[Style/ConstantNames] Constant name should be screaming-cased: LOG, not Log`

That's why I went against the specific `Log` naming convention initially but I'll change it.

I'll file an Ameba issue because other projects will face the same problem.

Also, I'm curious :
Is shainet used in production ? Do you (NeuraLegion) envision to make it grow ? Is it only maintained for educational purposes ?

File: shard.yml [hunk 14]
```diff
@@ -15,6 +15,6 @@ dependencies:
 development_dependencies:
   ameba:
     github: veelenga/ameba
-    version: 0.10.0
+    version: 0.13.0
```
Comment (Sija) [inline shard.yml:18]:
```suggestion
    version: ~> 0.13.0
```

File: src/shainet/basic/exprimental.cr (hunk 13 above)
Comment (rmarronnier) [reply src/shainet/basic/exprimental.cr]: I have no idea what the `#show` method does. 

Should I just remove the first argument or the method call altogether ?

File: shard.yml [hunk 15]
```diff
@@ -15,6 +15,6 @@ dependencies:
 development_dependencies:
   ameba:
     github: veelenga/ameba
-    version: 0.10.0
+    version:  ~> 0.13.0
```
Comment (Sija) [inline shard.yml:18]:
```suggestion
    version: ~> 0.13.0
```

File: src/shainet/basic/exprimental.cr [hunk 16]
```diff
@@ -39,7 +39,7 @@ module SHAInet
       end
 
       unless stealth # Hide report during training
-        @input_layers.last.activations.show(@logger, "input layer activations:")
+        @input_layers.last.activations.show(Log, "input layer activations:")
```
Comment (Sija) [inline src/shainet/basic/exprimental.cr:42]: As I said before, passing a `Log` instance shouldn't be needed.
ditto elsewhere

File: src/shainet/basic/exprimental.cr [hunk 17]
```diff
@@ -146,7 +146,7 @@ module SHAInet
       # This methods accepts data as either a SHAInet::TrainingData object, or as an Array(Array(Array(GenNum)).
       # In the case of SHAInet::TrainingData, we convert it to an Array(Array(Array(GenNum)) by calling #data on it.
       raw_data = data.is_a?(SHAInet::TrainingData) ? data.data : data
-      @logger.info("Training started")
+      Log.info("Training started")
```
Comment (Sija) [inline src/shainet/basic/exprimental.cr:149]:
```suggestion
      Log.info { "Training started" }
```

ditto elsewhere

File: src/shainet/cnn/fc_layer.cr [hunk 18]
```diff
@@ -1,15 +1,15 @@
-require "logger"
+require "log"
 
 module SHAInet
   class FullyConnectedLayer
+    Log = ::Log.for(self)
```
Comment (Sija) [inline src/shainet/cnn/fc_layer.cr:5]:
```suggestion
    Log = ::Log.for(self)

```

File: src/shainet/cnn/max_pool_layer.cr [hunk 19]
```diff
@@ -1,13 +1,13 @@
-require "logger"
+require "log"
 
 module SHAInet
   class MaxPoolLayer
+    Log = ::Log.for(self)
```
Comment (Sija) [inline src/shainet/cnn/max_pool_layer.cr:5]:
```suggestion
    Log = ::Log.for(self)

```

File: src/shainet/cnn/relu_layer.cr [hunk 20]
```diff
@@ -1,13 +1,13 @@
-require "logger"
+require "log"
 
 module SHAInet
   class ReluLayer
+    Log = ::Log.for(self)
```
Comment (Sija) [inline src/shainet/cnn/relu_layer.cr:5]:
```suggestion
    Log = ::Log.for(self)

```

File: src/shainet/cnn/softmax_layer.cr [hunk 21]
```diff
@@ -1,11 +1,12 @@
-require "logger"
+require "log"
 
 module SHAInet
   class SoftmaxLayer
+    Log = ::Log.for(self)
```
Comment (Sija) [inline src/shainet/cnn/softmax_layer.cr:5]:
```suggestion
    Log = ::Log.for(self)

```

File: src/shainet/data/json_data.cr [hunk 22]
```diff
@@ -2,34 +2,30 @@ require "json"
 
 module SHAInet
   class NetDump
-    JSON.mapping({
-      layers: Array(LayerDump),
-    })
+    include JSON::Serializable
```
Comment (Sija) [inline src/shainet/data/json_data.cr:5]:
```suggestion
    include JSON::Serializable

```

File: src/shainet/data/json_data.cr [hunk 23]
```diff
@@ -2,34 +2,30 @@ require "json"
...
+    property layers : Array(LayerDump)
   end
 
   class LayerDump
-    JSON.mapping({
-      l_type:              String,
-      neurons:             Array(NeuronDump),
-      activation_function: String,
-    })
+    include JSON::Serializable
```
Comment (Sija) [inline src/shainet/data/json_data.cr:10]:
```suggestion
    include JSON::Serializable

```

File: src/shainet/data/json_data.cr [hunk 24]
```diff
@@ -2,34 +2,30 @@ require "json"
...
 
   class NeuronDump
-    JSON.mapping({
-      id:           String,
-      bias:         Float64,
-      n_type:       String,
-      synapses_in:  Array(SynapseDump),
-      synapses_out: Array(SynapseDump),
-    })
+    include JSON::Serializable
```
Comment (Sija) [inline src/shainet/data/json_data.cr:17]:
```suggestion
    include JSON::Serializable

```

File: src/shainet/data/json_data.cr [hunk 25]
```diff
@@ -2,34 +2,30 @@ require "json"
...
+    property synapses_out : Array(SynapseDump)
   end
 
   class SynapseDump
-    JSON.mapping({
-      source:      String,
-      destination: String,
-      weight:      Float64,
-    })
+    include JSON::Serializable
```
Comment (Sija) [inline src/shainet/data/json_data.cr:26]:
```suggestion
    include JSON::Serializable

```

File: src/shainet/rnn/rnn.cr [hunk 26]
```diff
@@ -0,0 +1,367 @@
...
+
+module SHAInet
+  alias RNNLayer = InputLayer | ReluLayer | MaxPoolLayer | FullyConnectedLayer | DropoutLayer | SoftmaxLayer
+  alias RNNPair = {input: Array(Array(Array(Float64))), output: Array(Float64)}
+
+  # Note: Data is stored within specific classes.
+  # Structure hierarchy: RNN > Layer > Filter > Channel > Row > Neuron/Synapse
+
+  # Recurrent Neural Network
+  class RNN
```
Comment (Sija) [inline src/shainet/rnn/rnn.cr:12]: Why was this file added?

Comment (rmarronnier) [reply src/shainet/rnn/rnn.cr]: My bad, it's a unrelated WIP, it wasn't supposed to be commited in this PR. Sorry about that.

File: src/shainet/basic/exprimental.cr (hunk 13 above)
Comment (Sija) [reply src/shainet/basic/exprimental.cr]: I'd remove the argument.

File: src/shainet/basic/network_run.cr [hunk 27]
```diff
@@ -115,8 +115,8 @@ module SHAInet
       # This methods accepts data as either a SHAInet::TrainingData object, or as an Array(Array(Array(GenNum)).
       # In the case of SHAInet::TrainingData, we convert it to an Array(Array(Array(GenNum)) by calling #data on it.
       raw_data = data.is_a?(SHAInet::TrainingData) ? data.data : data
-      @logger.info("Training started")
-      start_time = Time.new
+      Log.info { "Training started" }
+      start_time = Time.local
```
Comment (Sija) [inline src/shainet/basic/network_run.cr:119]:
```suggestion
      start_time = Time.monotonic
```

File: src/shainet/basic/network_run.cr [hunk 28]
```diff
@@ -132,15 +132,15 @@ module SHAInet
...
+            # Log.info("Network saved.")
             save_to_file("#{autosave[:path]}/autosave_epoch_#{counter}.nn")
           end
         end
 
         # Break condtitions
         if counter >= epochs || (error_threshold >= @mse) && (counter > 1)
           log_summary(counter)
-          @logger.info("Training finished. (Elapsed: #{Time.new - start_time})")
+          Log.info { "Training finished. (Elapsed: #{Time.local - start_time})" }
```
Comment (Sija) [inline src/shainet/basic/network_run.cr:143]:
```suggestion
          Log.info { "Training finished. (Elapsed: #{Time.monotonic - start_time})" }
```

File: src/shainet/basic/network_run.cr [hunk 29]
```diff
@@ -390,8 +390,8 @@ module SHAInet
       # This methods accepts data as either a SHAInet::TrainingData object, or as an Array(Array(Array(GenNum)).
       # In the case of SHAInet::TrainingData, we convert it to an Array(Array(Array(GenNum)) by calling #data on it.
       raw_data = data.is_a?(SHAInet::TrainingData) ? data.data : data
-      @logger.info("Training started")
-      start_time = Time.new
+      Log.info { "Training started" }
+      start_time = Time.local
```
Comment (Sija) [inline src/shainet/basic/network_run.cr:394]:
```suggestion
      start_time = Time.monotonic
```

File: src/shainet/basic/network_run.cr [hunk 30]
```diff
@@ -406,15 +406,15 @@ module SHAInet
...
+            # Log.info("Network saved.")
             save_to_file("#{autosave[:path]}/autosave_epoch_#{epoch}.nn")
           end
         end
 
         # Break condtitions
         if epoch >= epochs || (error_threshold >= @mse) && (epoch > 1)
           log_summary(epoch)
-          @logger.info("Training finished. (Elapsed: #{Time.new - start_time})")
+          Log.info { "Training finished. (Elapsed: #{Time.local - start_time})" }
```
Comment (Sija) [inline src/shainet/basic/network_run.cr:417]:
```suggestion
          Log.info { "Training finished. (Elapsed: #{Time.monotonic - start_time})" }
```

File: src/shainet/basic/exprimental.cr [hunk 31]
```diff
@@ -146,7 +146,7 @@ module SHAInet
       # This methods accepts data as either a SHAInet::TrainingData object, or as an Array(Array(Array(GenNum)).
       # In the case of SHAInet::TrainingData, we convert it to an Array(Array(Array(GenNum)) by calling #data on it.
       raw_data = data.is_a?(SHAInet::TrainingData) ? data.data : data
-      @logger.info("Training started")
+      Log.info { "Training started" }
       start_time = Time.new
```
Comment (Sija) [inline src/shainet/basic/exprimental.cr:150]:
```suggestion
      start_time = Time.monotonic
```

File: src/shainet/basic/exprimental.cr [hunk 32]
```diff
@@ -162,15 +162,15 @@ module SHAInet
...
+            # Log.info("Network saved.")
             save_to_file("#{autosave[:path]}/autosave_epoch_#{counter}.nn")
           end
         end
 
         # Break condtitions
         if counter >= epochs || (error_threshold >= @mse) && (counter > 1)
           log_summary(counter)
-          @logger.info("Training finished. (Elapsed: #{Time.new - start_time})")
+          Log.info { "Training finished. (Elapsed: #{Time.new - start_time})" }
```
Comment (Sija) [inline src/shainet/basic/exprimental.cr:173]:
```suggestion
          Log.info { "Training finished. (Elapsed: #{Time.monotonic - start_time})" }
```

File: shard.yml (hunk 15 above)
Comment (Sija) [reply shard.yml:18]: it's not resolved, still

File: src/shainet/basic/network_setup.cr [hunk 33]
```diff
@@ -202,7 +204,7 @@ module SHAInet
           end
         end
       end
-      @logger.info("Cleaned #{current_neuron_number - @all_neurons.size} dead neurons")
+      Log.info("Cleaned #{current_neuron_number - @all_neurons.size} dead neurons")
```
Comment (Sija) [inline src/shainet/basic/network_setup.cr:207]:
```suggestion
      Log.info { "Cleaned #{current_neuron_number - @all_neurons.size} dead neurons" }
```

File: src/shainet/basic/exprimental.cr [hunk 34]
```diff
@@ -162,15 +162,15 @@ module SHAInet
         # Autosave the network
         unless autosave.nil?
           if counter % autosave[:freq] == 0 && (counter > 0)
-            # @logger.info("Network saved.")
+            # Log.info("Network saved.")
```
Comment (Sija) [inline src/shainet/basic/exprimental.cr:165]:
```suggestion
            # Log.info { "Network saved." }
```

File: src/shainet/basic/exprimental.cr [hunk 35]
```diff
@@ -223,8 +223,8 @@ module SHAInet
           # Show training progress of the mini-batches
           i += 1
           if counter % log_each == 0
-            @logger.info("  Slice: (#{i} / #{slices}), MSE: #{@mse}") if show_slice
-            # @logger.info("@error_signal: #{@error_signal}")
+            Log.info { "  Slice: (#{i} / #{slices}), MSE: #{@mse}" } if show_slice
+            # Log.info("@error_signal: #{@error_signal}")
```
Comment (Sija) [inline src/shainet/basic/exprimental.cr:227]:
```suggestion
            # Log.info { "@error_signal: #{@error_signal}" }
```

File: src/shainet/basic/network_run.cr [hunk 36]
```diff
@@ -132,15 +132,15 @@ module SHAInet
         # Autosave the network
         unless autosave.nil?
           if counter % autosave[:freq] == 0 && (counter > 0)
-            # @logger.info("Network saved.")
+            # Log.info("Network saved.")
```
Comment (Sija) [inline src/shainet/basic/network_run.cr:135]:
```suggestion
            # Log.info { "Network saved." }
```

File: src/shainet/basic/network_run.cr [hunk 37]
```diff
@@ -235,8 +235,8 @@ module SHAInet
           # Show training progress of the mini-batches
           display_counter += 1
           if counter % log_each == 0
-            @logger.info("  Slice: (#{display_counter} / #{slices}), MSE: #{@mse}") if show_slice
-            # @logger.info("@error_signal: #{@error_signal}")
+            Log.info { "  Slice: (#{display_counter} / #{slices}), MSE: #{@mse}" } if show_slice
+            # Log.info("@error_signal: #{@error_signal}")
```
Comment (Sija) [inline src/shainet/basic/network_run.cr:239]:
```suggestion
            # Log.info { "@error_signal: #{@error_signal}" }
```

File: src/shainet/basic/network_run.cr [hunk 38]
```diff
@@ -406,15 +406,15 @@ module SHAInet
         # Autosave the network
         unless autosave.nil?
           if epoch % autosave[:freq] == 0 && (epoch > 0)
-            # @logger.info("Network saved.")
+            # Log.info("Network saved.")
```
Comment (Sija) [inline src/shainet/basic/network_run.cr:409]:
```suggestion
            # Log.info { "Network saved." }
```

File: src/shainet/basic/network_run.cr [hunk 39]
```diff
@@ -470,8 +470,8 @@ module SHAInet
           # Show training progress of the mini-batches
           display_counter += 1
           if epoch % log_each == 0
-            @logger.info("  Slice: (#{display_counter} / #{slices}), MSE: #{@mse}") if show_slice
-            # @logger.info("@error_signal: #{@error_signal}")
+            Log.info { "  Slice: (#{display_counter} / #{slices}), MSE: #{@mse}" } if show_slice
+            # Log.info("@error_signal: #{@error_signal}")
```
Comment (Sija) [inline src/shainet/basic/network_run.cr:474]:
```suggestion
            # Log.info { "@error_signal: #{@error_signal}" }
```

File: src/shainet/basic/exprimental.cr [hunk 40]
```diff
@@ -82,14 +82,14 @@ module SHAInet
       # msg += "\n  Input:\n  #{input_data}"
       # msg += "\n  Output:\n  #{actual_output}"
       # msg += "\n  Expected:\n  #{expected_output}"
-      # @logger.info(msg)
+      # Log.info(msg)
```
Comment (Sija) [inline src/shainet/basic/exprimental.cr:85]:
```suggestion
      # Log.info { msg }
```

File: src/shainet/basic/exprimental.cr [hunk 41]
```diff
@@ -82,14 +82,14 @@ module SHAInet
...
       # end
 
       # Detect exploading gradiants and NaNs in output
       validate_values(actual_output, "actual_output")
       # actual_output.each do |ar|
       #   if ar.infinite?
-      #     @logger.info("Found an '#{ar}' value, run stopped.")
+      #     Log.info("Found an '#{ar}' value, run stopped.")
```
Comment (Sija) [inline src/shainet/basic/exprimental.cr:92]:
```suggestion
      #     Log.info { "Found an '#{ar}' value, run stopped." }
```

**Comment by ArtLinkov** [comment]
@rmarronnier thanks for the PR, it is greatly appreciated :slightly_smiling_face: 

> Also, I'm curious :
Is shainet used in production ? Do you (NeuraLegion) envision to make it grow ? Is it only maintained for educational purposes ?

Originally SHAInet was developed as a research and educational project, and for now we don't have any immediate plans to add additional production optimizations (like MT support, etc.). However, I do have some idea's in the back of my mind to add some cool features eventually, just waiting for some free time on my hands. 

In regards to the growth of the project, absolutely! Right now it's mostly a community effort, we embrace all contributions :slightly_smiling_face:, but we do plan to allocate resources back to this project in the future

**Comment by rmarronnier** [comment]
@ArtLinkov Thanks for the comprehensive answer :-)

I'm not sure I can contribute much to Shainet (low level ML is out of my league). I might try though ;-)

However, I keep a tab on it hoping I can use it one day to implement some state of the art NLP processes for Cadmium.

**Comment by ArtLinkov** [comment]
Looks good, merging.
@Sija thanks for the review :slightly_smiling_face:
